src/coreComponents/mesh/generators/CellBlockABC.hpp

```diff
@@ -0,0 +1,131 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2020-     GEOSX Contributors
+ * All right reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+#ifndef GEOSX_CELLBLOCKABC_HPP
+#define GEOSX_CELLBLOCKABC_HPP
+
+#include "dataRepository/Group.hpp"
+#include "common/DataTypes.hpp"
+
+#include <vector>
+
+namespace geosx
+{
+
+/**
+ * Abstract base class defining the information provided by any cell block implementation.
+ * Mainly element type related information (number of faces per element...),
+ * some element to nodes, faces and edges mappings, a local to global mapping and
+ * some kind of accessors to external properties.
+ *
+ * The cells of CellBlockABC are all of the same kind.
+ *
+ * It's noteworthy that the CellBlockABC is immutable oriented.
+ * The derived implementations need to have the modification/creation capabilities.
+ */
+class CellBlockABC : public dataRepository::Group
+{
+public:
+
+  /**
+   * @brief Constructor
+   * @param name The name of this Group.
+   * @param parent The parent Group.
+   */
+  CellBlockABC( string const & name,
+                Group * const parent )
+    :
+    Group( name, parent )
+  { }
+
+  /**
+   * @brief Get the type of element in this subregion.
+   * @return a string specifying the type of element in this subregion
+   *
+   * See class FiniteElementBase for possible element type.
+   */
+  virtual string getElementTypeString() const = 0;
+
+  /**
+   * @brief Get the number of nodes per element.
+   * @return number of nodes per element
+   */
+  virtual localIndex numNodesPerElement() const = 0;
+
+  /**
+   * @brief Get the number of faces per element.
+   * @return number of faces per element
+   */
+  virtual localIndex numFacesPerElement() const = 0;
+
+  /**
+   * @brief Get the number of elements.
+   * @return number of elements in the cell block
+   */
+  virtual localIndex numElements() const = 0;
+
+  /**
+   * @brief Get the element-to-node map.
+   * @return a reference to the element-to-node map
+   */
+  virtual array2d< localIndex, cells::NODE_MAP_PERMUTATION > const & getElemToNode() const = 0;
```

Comment:
This contract seems to bind all implementations of `CellBlockABC` to a specific elem-to-node layout that is `array2d`. However "`CrazyCellBlock`" will need to use something like `ArrayOfArrays` to store its varying per-cell topology (same goes for other maps in this class), and trying to fit it into an `array2d` would be extremely memory-inefficient (think of one crazy cell that has 50 nodes in a block of otherwise "mildly" unstructured cells).

Two approaches come to mind:
1. Make the return type here the most flexible one (`ArrayOfArrays`) and make "structured" subregions like do the legwork to convert it into their internal storage format (`array2d`). This is runtime overhead, but maybe it's acceptable if it happens once per simulation.
2. Make two virtual functions (`getElemToNodeUniform()` and `getElemToNodeNonUniform()`) with different return types plus a discriminator (`bool isUniform()`), such that existing `CellElementSubRegion` will only call the former and new `CrazyElementSubRegion` will call the latter. This partially defeats the purpose of an abstract base class (?), but allows each subregion to call the implementation that is most efficient. 

Maybe my comment jumps ahead a bit, because we don't have `CrazyCellBlock` yet, and it's not decided how it's topology will be represented and stored in a subregion (I just assumed it will be `ArrayOfArrays`, but who knows). But since this whole refactoring is motivated by the need to prepare for unstructured mesh readers, I find it useful to think in advance what it's going to look like and how it can implement the ABC.

Reply:
Hi @klevzoff thx for the comments, I appreciate.

You’re right, I split the code and kept the types as they are right now.
This is indeed questionable for the long term.
There are three elements I’d like to mention as an answer.

- First, I noticed that there are some extra allocations here and there (see the usage of `CellBlockManagerABC::getEdgeMapOverallocation` alike function for example).
I like `CellBlock*` being considered as a raw geometric part of the code; thus I do not know if this is the proper location for having this extra allocations for computations.
(Also, we could define a policy about this overallocation in the input file not to impose the extra allocation for solvers which do not need it. But where in the code?)
This is somehow related to your point: how can we know which memory layout is the best for client code? There is surely a way, but what would be the cost ? (ease of use, readibility, maintenance, performance…)

- Second, imho the root cause of the issue you are pointing is that we are somehow mixing concepts (`CellBlockABC`, `CellBlockManagerABC`) and implementations (`array2d`, `ArraysOfArrays`) here.
For example `CellBLockABC::getElemToEdges` returns an `array2d< localIndex >` but the concept behind this “_matrix_” is in fact an “_element to edges mapping, as a matter of fact implemented under the form of an `array2d< localIndex >`_”.
In an ideal world we could use some kind of our `FixedOneToManyRelation` or `OrderedVariableToManyElementRelation`… which would be abstract too. The actual underlying storage could possibly be prescribed based on the solver, the user... If performances are at stake, maybe down casting the underlying data structure could be considered?

- Last but important point related to your architecture concern.
So far my split binds to GEOSX’s capabilities, which is `CellElementSubRegion` has one unique type of cells.
When we have crazy cells, we probably have to transform some of the core and solvers of GEOSX too; and the `CellBlock*ABC` will follow.
Maybe will we have `Heterogeneous` and `Homogeneous` `CellElementSubRegions`? _I do not know yet, that’s the critical upcoming question_.
And when this question is answered, I’m convinced that `CellBlock*ABC` will follow the same path consistently. Look: `CellBlockABC::numFacesPerElement()` has no meaning for non homogeneous sub-regions. This is even worse 😨 😱 

And [as Joel says](https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/) “_All non-trivial abstractions, to some degree, are leaky_.” We have to find the best balance!